jupyter_kernel_mgmt/subproc/launcher.py

```diff
@@ -138,18 +140,29 @@ def format_kernel_cmd(self, connection_file, kernel_resource_dir=None):
             # but it should be.
             cmd[0] = sys.executable
 
-        ns = dict(connection_file=connection_file,
+        # Preserve system-owned substitutions by starting with launch params
+        ns = dict()
+        if isinstance(self.launch_params, dict):
+            ns.update(self.launch_params)
```

Comment:
Do we want to drop all launch_params into this namespace, or have a sub-dict for that and leave room in case there's a need for other kinds of parameters in the future?

Reply:
I think I see your point relative to another area I was thinking about - environment variables - where the metadata also includes a list of supported environment variables.  We could then maintain separate dicts.  That said, having everything in the same namespace is only an issue if there are conflicts between env names and launch args and I'd be okay with stating that those spaces are shared.

I think we could keep a single set of parameters by adding meta-property indicating their expected usages.  For example, a `context` meta-property could be included that has one of a set of values like {'env', 'argv', 'config'}.  The presentation layer could choose to use the context to split how it prompts for the values or not - its up to them.  The provider could decompose list into multiple sets as it chooses.

I think this comes down to extensibility and my feeling is it would be better to extend the schema with an additional context type vs. supporting a varying number of sub-dicts - where each sub-dict infers its context.

That said, my preference toward a metadata approach is just slightly stronger.